### PR TITLE
UMD support

### PR DESCRIPTION
**Pull request by zewa666**
Adds UMD support for the libraries build artifacts

Related issue: https://github.com/chartjs/chartjs-plugin-deferred/issues/2

### CODE REVIEW

**Comment by zewa666** [comment]
As explained in the referenced issue I wouldn't see any easy way, for any module bundler to take full control, over when dependencies are properly timed. So an UMD approach seems to be the best to satisfy a broad variety of module formats.

Please do make sure this change doesn't break any pre-existing use cases, I didn't really find anything special but it would make sense to take some existing examples and properly check those.

I've also seen that you leverage gulptasks, so I felt free to modify those, but you might wanna think about creating a 3rd build articaft containing the umd exports

File: src/plugin.js [hunk 1]
```diff
@@ -1,206 +1,202 @@
...
-		}
-	});
-
-	function computeOffset(value, base) {
-		var number = parseInt(value, 10);
-		if (isNaN(number)) {
-			return 0;
-		} else if (typeof value === 'string' && value.indexOf('%') !== -1) {
-			return number / 100 * base;
+var Chart = ChartJS || window.Chart;
```
Comment (simonbrunel) [inline src/plugin.js]: Changing the whole file indentation by removing the surrounding function make the PR almost unreadable and will inevitably conflict with the upcoming changes I'm working on based on Chart.js 2.6. Also, I would like to keep that file directly injectable (`<script>`) and thus preserve the global function scoping. That's why I prefer the following approach (inject the UMD param as function parameter):

```javascript
/* global window: false, Chart */

'use strict';

(function(Chart) {

    // remove: var Chart = window.Chart;
    // and keep previous code unchanged
    // ...

})(Chart || window.Chart);
```

File: gulpfile.js [hunk 2]
```diff
@@ -12,6 +12,7 @@ var zip = require('gulp-zip');
 var merge = require('merge2');
 var path = require('path');
 var package = require('./package.json');
+var umd = require('gulp-umd');
```
Comment (simonbrunel) [inline gulpfile.js]: Can you move this require right after the `var uglify = require('gulp-uglify');` one?

File: gulpfile.js [hunk 3]
```diff
@@ -46,6 +47,19 @@ function watch(glob, task) {
...
     return gulp.src(srcDir + 'plugin.js')
+      .pipe(umd({
+        dependencies: function(file) {
+          return [
+            {
+              name: 'chart.js',
+              amd: 'chart.js',
+              cjs: 'chart.js',
+              global: 'Chart',
+              param: 'ChartJS'
```
Comment (simonbrunel) [inline gulpfile.js]: Can the UMD param be renamed to `Chart` and reformat the array as:

```javascript
return [{
  name: 'chart.js',
  amd: 'chart.js',
  cjs: 'chart.js',
  global: 'Chart',
  param: 'Chart'
}]
```

**Comment by zewa666** [comment]
Alright I've added the requested changes. One tip. How about rewriting the gulpfile to ES6 so it could be removed from the eslintignore list?

**Comment by simonbrunel** [comment]
Looks good, thank, will try to test that very soon! The `gulpfile` is in the `eslintignore` because it doesn't use the same indentation rules as the plugin. Though, I already have some changes locally and can have a look to migrate to ES6.

That could be interesting to look at the other Chart.js hosted plugins to make them fully compatible with other environments such as Aurelia ;)

File: src/plugin.js [hunk 4]
```diff
@@ -1,4 +1,4 @@
-/* global window: false, ChartJS */
```
Comment (simonbrunel) [inline src/plugin.js]: Actually, `Chart` needs to be defined in the global (see Travis build)

**Comment by zewa666** [comment]
@simonbrunel wonder why my local eslint didn't capture that ... strange. Good catch thanks

**Comment by simonbrunel** [comment]
`gulp-umd` uses the filename (capitalized) as the global namespace and so tries to set `window.Plugin = Plugin`, which doesn't look nice. Also, the plugin doesn't export anything and `module.exports = undefined` doesn't look good either.

It's possible to configure the target [`namespace`](https://github.com/eduardolundgren/gulp-umd#options) but since `gulp-umd` doesn't do a lot and seems not anymore maintained, I think it's easier to directly implement UMD inside `plugin.js`:

```javascript
'use strict';

;(function(root, factory) {
	if (typeof define === 'function' && define.amd) {
		define(['chart.js'], factory);
	} else if (typeof exports === 'object') {
		module.exports = factory(require('chart.js'));
	} else {
		factory(window.Chart);
	}
}(this, function(Chart) {
	var helpers = Chart.helpers;

        // ...

	var plugin = {
		beforeInit: function(instance) {
		// ...
	};

	Chart.plugins.register(plugin);
	return plugin;
}));
```

**Comment by zewa666** [comment]
Fewer dependencies is always better. I was just unsure whether you'd like to have it as part of your gulp build, but if you're fine with placing it in plugin.js it's definitely fine as well

**Comment by simonbrunel** [comment]
@zewa666 I'm fine with having it in `plugin.js`, do you want to update that PR or should I implement it in my upcoming changes?

**Comment by zewa666** [comment]
hey @simonbrunel, yeah if you don't mind that would be easier. I'll close this PR thanks for the reviews. 👍

**Comment by simonbrunel** [comment]
No problem, thanks for your help and time :)

**Comment by simonbrunel** [comment]
After investigating a bit more, I think I will use [rollup](`rollup/rollup`) which generates a better UMD code than `gulp-umd`.

**Comment by zewa666** [comment]
Sure, which also has a lot of traction these days. Good choice

**Pull request closed without merging**